src/dex/spectrum.ts

```diff
@@ -0,0 +1,209 @@
+import { LiquidityPool } from './models/liquidity-pool';
+import { BaseDataProvider } from '@providers/data/base-data-provider';
+import { Asset, Token } from './models/asset';
+import { BaseDex } from './base-dex';
+import { DatumParameters, DefinitionConstr, DefinitionField, PayToAddress, RequestConfig, SwapFee, UTxO } from '@app/types';
+import { DefinitionBuilder } from '@app/definition-builder';
+import { AddressType, DatumParameterKey } from '@app/constants';
+import { BaseApi } from '@dex/api/base-api';
+import pool from './definitions/spectrum/pool';
+import order from './definitions/spectrum/order';
+import { all, bignumber, BigNumber, ConfigOptions, create, FormatOptions, MathJsStatic } from 'mathjs';
+
+export class Spectrum extends BaseDex {
+  public static readonly identifier: string = 'Spectrum';
+  public readonly api: BaseApi;
+
+  /**
+   * On-Chain constants.
+   */
+  public readonly orderAddress: string = 'addr1wynp362vmvr8jtc946d3a3utqgclfdl5y9d3kn849e359hsskr20n';
+  public readonly poolAddress: string = 'addr1x8nz307k3sr60gu0e47cmajssy4fmld7u493a4xztjrll0aj764lvrxdayh2ux30fl0ktuh27csgmpevdu89jlxppvrswgxsta';
+
+  constructor(requestConfig: RequestConfig = {}) {
+    super();
+  }
+  public async liquidityPoolAddresses(provider: BaseDataProvider): Promise<string[]> {
+    return Promise.resolve([this.poolAddress]);
+  }
+
+  async liquidityPools(provider: BaseDataProvider): Promise<LiquidityPool[]> {
+    const utxos: UTxO[] = await provider.utxos(this.poolAddress);
+
+    return await Promise.all(
+      utxos.map(async (utxo: UTxO) => {
+        return await this.liquidityPoolFromUtxo(provider, utxo);
+      })
+    ).then((liquidityPools: (LiquidityPool | undefined)[]) => {
+      return liquidityPools.filter((liquidityPool?: LiquidityPool) => {
+        return liquidityPool !== undefined;
+      }) as LiquidityPool[];
+    });
+  }
+
+  public async liquidityPoolFromUtxo(provider: BaseDataProvider, utxo: UTxO): Promise<LiquidityPool | undefined> {
+    if (!utxo.datumHash) {
+      return Promise.resolve(undefined);
+    }
+
+    const relevantAssets = utxo.assetBalances.filter((assetBalance) => {
+      const assetName = (assetBalance.asset as Asset).assetName;
+      return !assetName?.toLowerCase()?.endsWith('_nft') && !assetName?.toLowerCase()?.endsWith('_lq'); // TODO Improve method of identification here to not rely solely on an assetName.
+    });
```

Comment:
Since the order datum contains the swap tokens info, we can use those to grab the relevant assets. Im sure this method works, but using the datum parameters is a little safer